server.py

```diff
@@ -81,37 +74,24 @@ def signal_handler(sig: int, _frame):
         service.initialize() for service in services.values()
     ])
 
-    if config.PROFILING_INTERVAL > 0:
-        logger.warning("Profiling enabled! This will create additional load.")
-        import cProfile
-        pr = cProfile.Profile()
-        profiled_count = 0
-        max_count = 300
-
-        @at_interval(config.PROFILING_INTERVAL, loop=loop)
-        async def run_profiler():
-            nonlocal profiled_count
-            nonlocal pr
-
-            if len(services["player_service"]) > 1000:
-                return
-            elif profiled_count >= max_count:
-                del pr
-                return
-
-            logger.info("Starting profiler")
-            pr.enable()
-            await asyncio.sleep(2)
-            pr.disable()
-            profiled_count += 1
-
-            logging.info("Done profiling %i/%i", profiled_count, max_count)
-            pr.dump_stats("profile.txt")
+    make_profiler = get_profiler_factory(services["player_service"])
+    await make_profiler()
+    config.register_callback("PROFILING_INTERVAL", make_profiler)
 
     ctrl_server = await server.run_control_server(
         services["player_service"],
         services["game_service"]
     )
+    async def restart_control_server():
+        nonlocal ctrl_server
+        nonlocal services
+
+        ctrl_server.shutdown()
```

Comment:
shutdown is a coroutine
```suggestion
        await ctrl_server.shutdown()
```